### PR TITLE
chore: consolidated API: add anonymous user check

### LINKED ISSUE #30063: [Task]: Consolidated API : add authentication check to mimic original API endpoints
### SubTasks

- In case original API returned 401 status for un-authenticated requests, then the sub API in consolidated API should also return the same status.

### PR DESCRIPTION
**Pull request by sumitsum**
## Description
- Check if sub API is accessible by an anonymous user or not. If not, then return 401 error. This is done to mimic the behaviour by individual APIs where they return 401 error if they are not accessible by an anonymous user. 
- The root API status is not impacted by this. 
- This is targeted to merge in dev branch `oneApiChanges`

#### PR fixes following issue(s)
Fixes #30063 

#### Media
Screenshot 2024-01-08 at 8 17 19 AM

#### Type of change
- Chore (housekeeping or task changes that don't impact user perception)

#### How Has This Been Tested?
- [x] Manual
- For this the same JUnit TC should suffice that needs to be written to check for response against anonymous user. It is tracked [here](https://github.com/appsmithorg/appsmith/issues/29988). I have skipped writing JUnit TC because currently the test suite is disabled due to error on EE. 

#### Dev activity
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] PR is being merged under a feature flag

#### QA activity:
- [ ] [Speedbreak features](https://github.com/appsmithorg/TestSmith/wiki/Guidelines-for-test-plans#speedbreakers-) have been covered
- [ ] Test plan covers all impacted features and [areas of interest](https://github.com/appsmithorg/TestSmith/wiki/Guidelines-for-test-plans#areas-of-interest-)
- [ ] Test plan has been peer reviewed by project stakeholders and other QA members
- [ ] Manually tested functionality on DP
- [ ] We had an implementation alignment call with stakeholders post QA Round 2
- [ ] Cypress test cases have been added and approved by SDET/manual QA
- [ ] Added `Test Plan Approved` label after Cypress tests were reviewed
- [ ] Added `Test Plan Approved` label after JUnit tests were reviewed

### CODE REVIEW

File: app/server/appsmith-server/src/main/java/com/appsmith/server/services/ConsolidatedAPIServiceImpl.java [hunk 1]
```diff
@@ -131,6 +152,20 @@ <T> Mono<ResponseDTO<T>> getErrorResponseMono(Throwable error, Class<T> type) {
                 INTERNAL_SERVER_ERROR_STATUS, new ErrorDTO(INTERNAL_SERVER_ERROR_CODE, error.getMessage())));
     }
 
+    /**
+     * This method checks if a given sub API is accessible to the anonymous user or not. If not, then it returns 401
+     * error. Whether a sub API is accessible or not is given by the map IS_API_ACCESSIBLE_TO_ANONYMOUS_USER_MAP .
+     */
+    Mono<Boolean> checkApiAccessIfAnonymousUser(Mono<User> userMono, String api) {
+        return userMono.flatMap(user -> {
```
Comment (rajatagrawal) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/services/ConsolidatedAPIServiceImpl.java:160]: Can we have an integration unit test for this feature ?

File: app/server/appsmith-server/src/main/java/com/appsmith/server/services/ConsolidatedAPIServiceImpl.java [hunk 2]
```diff
@@ -131,6 +152,20 @@ <T> Mono<ResponseDTO<T>> getErrorResponseMono(Throwable error, Class<T> type) {
...
+            if (!user.getIsAnonymous() || IS_API_ACCESSIBLE_TO_ANONYMOUS_USER_MAP.get(api)) {
```
Comment (rajatagrawal) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/services/ConsolidatedAPIServiceImpl.java:161]: I personally find it hard to understand the if statement with a negation and a logically or. Would it be possible to rephrase the logic in a more understandable way ? For example:

```
if (user.isAnonymous()) {
  if (is_api_accessible_to_user(api)) {
    return true;
  } else {
    return exception;
  }
} else {
   return true;
}
```

File: app/server/appsmith-server/src/main/java/com/appsmith/server/services/ConsolidatedAPIServiceImpl.java (hunk 1 above)
Comment (sumitsum) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/services/ConsolidatedAPIServiceImpl.java:160]: Hi @rajatagrawal currently all tests are disabled due to the earler TBP issue on EE. Can we pick this up when the tests are enabled ?

**Review by rajatagrawal — CHANGES_REQUESTED** [review]
Unit tests are missing. How do ensure the authentication checks are working fine ?

File: app/server/appsmith-server/src/main/java/com/appsmith/server/services/ConsolidatedAPIServiceImpl.java (hunk 1 above)
Comment (rajatagrawal) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/services/ConsolidatedAPIServiceImpl.java:160]: Stability and quality of the code in prod is at your prerogative. Please use your best judgement and planning for it.

File: app/server/appsmith-server/src/main/java/com/appsmith/server/services/ConsolidatedAPIServiceImpl.java (hunk 2 above)
Comment (sumitsum) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/services/ConsolidatedAPIServiceImpl.java]: Hi @rajatagrawal re-factored the method as per suggestion.

Comment (rajatagrawal) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/services/ConsolidatedAPIServiceImpl.java]: Thanks!

**Comment by rajatagrawal** [comment]
Can you please recheck the branch history ? The number of commits is too high.

Also a suggestion,

Please prefer to merge all the branches to release and then frontend take a pull of it as a workflow.

The current workflow is erroneous and hard to keep a track of the changes.

File: app/server/appsmith-server/src/main/java/com/appsmith/server/services/ConsolidatedAPIServiceImpl.java [hunk 3]
```diff
@@ -130,6 +152,24 @@ <T> Mono<ResponseDTO<T>> getErrorResponseMono(Throwable error, Class<T> type) {
                 INTERNAL_SERVER_ERROR_STATUS, new ErrorDTO(INTERNAL_SERVER_ERROR_CODE, error.getMessage())));
     }
 
+    /**
+     * This method checks if a given sub API is accessible to the anonymous user or not. If not, then it returns 401
+     * error. Whether a sub API is accessible or not is given by the map IS_API_ACCESSIBLE_TO_ANONYMOUS_USER_MAP .
+     */
+    Mono<Boolean> checkApiAccessIfAnonymousUser(Mono<User> userMono, String api) {
```
Comment (mohanarpit) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/services/ConsolidatedAPIServiceImpl.java:159]: @sumitsum Based on my underestanding of the problem statement, I'm not sure if this is the best solution available to us. I'm wondering if we can do either of the following: 
1. Split the consolidated API into 2 parts: public v/s non-public endpoints. 
2. Split the consolidated API into 2 parts: edit v/s view mode. The edit mode will be authenticated and view mode will be public. 
3. Only implement for view mode.

If required, I'm happy to talk in a sync fashion as well about this.

Comment (sumitsum) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/services/ConsolidatedAPIServiceImpl.java:159]: cc: @rajatagrawal

**Comment by sumitsum** [comment]
As discussed, closing [this](https://theappsmith.slack.com/archives/C024GUDM0LT/p1704698113684959?thread_ts=1703074675.245879&cid=C024GUDM0LT) PR in favour of different endpoints for view and edit mode

**Pull request closed without merging**